66 model pipeline

doug-leasure opened the pull request:
This PR resolves issue #66. 

- Random effects on `p_F` and `p_G`
- Posterior predictive checks, including plots
- Improved plotting for model evaluation (trace plots, times series plots, etc)
- Support for running model and model evaluation from the command line
- Harmonise path structure with data pipeline
- Update data pipeline to allow locations to be excluded from analysis

doug-leasure committed:
model pipeline from command line

doug-leasure committed:
harmonise path definitions with data pipeline

doug-leasure committed:
Use ggplot2::ggsave to write trace plots to disk

doug-leasure committed:
Merge to resolve conflict

doug-leasure committed:
fix broken path, oops.

doug-leasure committed:
improve time series plotting function

doug-leasure committed:
add growth rates (r) to time series plots

doug-leasure committed:
delete ./.idea/

doug-leasure committed:
allow locations to be excluded from analysis

doug-leasure committed:
random effects on p_F and p_G; simplified r with informative prior

doug-leasure committed:
Merge branch '66-model-pipeline' of github.com:OxfordDemSci/UkraineNowPop into 66-model-pipeline

doug-leasure committed:
fix my bug in processed sma data

doug-leasure committed:
fix broken paths

doug-leasure committed:
in-sample posterior predictive checks

doug-leasure committed:
simplify setwd

doug-leasure commented:
Note:  This base model in the PR converges and gets relatively good fit with no covariates. The model is now ready for covariates on growth rates (r) and user ratios (p).

**In-sample posterior predictive check**
![image](https://github.com/user-attachments/assets/292cee12-58b0-4c02-bce8-f34e513028d4)

**Example estimates for Kyiv**
![image](https://github.com/user-attachments/assets/4eb6425e-1a9f-4d48-936e-b5fa68d674c0)

**Example trace plots for last time step in Kyiv**
![image](https://github.com/user-attachments/assets/31f04b3a-36d5-45da-8085-8eaf74323c2d)

doug-leasure committed:
reorganise modelling folders

doug-leasure commented:
FYI - I just noticed that I actually modelled `p_F` and `p_G` here as independent location- or time-specific fixed-effects. I will look into revising that before merging the PR, although I think it is okay to go ahead and review as is.

alaparicioc commented on src/1_pop_data/20_prepare_sma_data.R:
```diff
@@ -56,13 +56,17 @@ master_index_without_t <- master_index |>
 
 process_sma_data <- function(sma_data, metric = audience_metric) {
```
I cannot run from this line onwards, since the line "sma_facebook <- retrieve_sma_data("facebook")" produced "Error in as.data.frame.default(value, stringsAsFactors = FALSE) : 
  cannot coerce class ‘"function"’ to a data.frame"

doug-leasure replied:
@alaparicioc, I was able to replicate this error, but only when I had not successfully run the preceeding steps in the pipeline. When I run the previous steps successfully, this script works for me.

alaparicioc replied:
everything runs up to line 46 of https://github.com/OxfordDemSci/UkraineNowPop/blob/d63f0950e9971b87f9f889e339336b242d89c244/src/1_pop_data/20_prepare_sma_data.R

alaparicioc commented:
In “data_querying.R”, could something like the following be added?

package_name <- c(“jsonlite”, “httr”, “DBI”, “RPostgres”)
if (!require(package_name, character.only = TRUE)) { install.packages(package_name) library(package_name, character.only = TRUE) }

doug-leasure commented:
> In “data_querying.R”, could something like the following be added?
> 
> package_name <- c(“jsonlite”, “httr”, “DBI”, “RPostgres”) if (!require(package_name, character.only = TRUE)) { install.packages(package_name) library(package_name, character.only = TRUE) }

@alaparicioc I think this may already be taken care of here: 
https://github.com/OxfordDemSci/UkraineNowPop/blob/66-model-pipeline/requirements.R

doug-leasure committed:
proper random effects on p_F and p_G

edarin commented on example.env:
```diff
@@ -20,8 +20,9 @@ UN_STAFF_USERNAME=
 ACLED_EMAIL=
-in_dir="K:/DemSci/projects/2023_WHO_Ukraine_Population/input"
-out_dir="K:/DemSci/projects/2023_WHO_Ukraine_Population/output"
+repo_dir="~/git/OxfordDemSci/UkraineNowPop"
```
Why did you change the in_dir and out_dir?

doug-leasure replied:
I am happy to change this back, if needed. You can keep it however you want locally, though, so hopefully not a problem. 

I just wanted the example to indicate that it is possible to use the `wd` folder within the repository. It is included in the `.gitignore` for exactly this purpose, but I thought that may not be obvious to everyone. 

The other thought was that using the network drive is not going to be easy from inara, which is where much of our code will be run from.

edarin replied:
Fair.
Currently we have to be cautious as we are building as we go the inputs. And I think "K:/DemSci/projects/2023_WHO_Ukraine_Population/input" should be our internal reference. But I think we have it set it up locally like that/are aware of that so no problem.

edarin commented on src/1_pop_data/20_prepare_sma_data.R:
```diff
@@ -56,13 +56,17 @@ master_index_without_t <- master_index |>
     ) |>
     arrange(t_name) |>
-    mutate(m = format(t_name, "%u")) |>
+    mutate(m = format(collection_date, "%u")) |>
```
it seems weird as t_name is a weekly indexing when m is a daily indexing. it seems to me that it should be collection_date. But I might be missing something...

doug-leasure replied:
Here, I have switched to "collection_date" so that m aligns to individual days of the week. The previous code was assigning m=1 for all days of the week because it was using t_name.

edarin replied:
oh of course. I thought it was the opposite. I was wrong at the start. Perfect

edarin commented on src/1_pop_data/query_meta_keys.py:
```diff
@@ -27,8 +27,10 @@
 geo_level = 'regions'
 location_type = '["recent"]'
 
+drop_keys = ["3782", "3788", "3791", "3797"]
```
great use of the script! i shall rerun on the network drive for everyone to have the correct meta_keys.

doug-leasure commented:
@edarin , thanks for the nice review. I see you had requested changes, but I think that was resolved with your PR #69. Is that correct?

doug-leasure committed:
constrain priors on sigmas a little bit

alaparicioc commented:
Pipeline worked for me. 
About the results, I do not get convergence. There are 2,264 parameters that do not converge for me, which includes population parameters.
How many iterations did you use to get convergence? I used the parameters in the 1_mcmc.R file on https://github.com/OxfordDemSci/UkraineNowPop/blob/d63f0950e9971b87f9f889e339336b242d89c244/src/2_model/1_mcmc.R
chains <- 4
warmup <- 5e3
samples <- 5e3
inits <- lapply(1:chains, function(id) init_generator(md = md, chain_id = id))

In addition, I am getting very different results:
   
![image](https://github.com/user-attachments/assets/24a12ac0-dbde-4809-bd71-2841737dc1f5)

![image](https://github.com/user-attachments/assets/fe72d006-edbf-46f1-aa73-1efd03093ca4)

![image](https://github.com/user-attachments/assets/8562b0e5-1d6a-48e3-be55-b2d3939fa807)

What are the summary statistics of GF ratio you include as data in the model?

doug-leasure commented:
@alaparicioc, that is great that you got the entire pipeline to run! 

The code that generates the FG ratio data is [here](https://github.com/OxfordDemSci/UkraineNowPop/blob/66-model-pipeline/src/2_model/models/1_base_model_config.R#L61).

I got fairly good convergence with 5,000 burnin and 5,000 samples which took about 75 minute to run. I think all parameters had rhat < 1.1, but would need to double check to be sure. 

Have you have included Luhansk, Donetsk, Crimea, and Sevastopol in your model? The data in those oblasts causes problems that I think will be best resolved by including covariates in the model. For the results that I showed, I removed those oblasts for now.

Can you check if you have those oblasts in your model?

If so, it is because the `.../inputs/meta_keys.csv` file on the DemSci drive still included those oblasts (I regenerated that file locally using [query_meta_keys.py](https://github.com/OxfordDemSci/UkraineNowPop/blob/66-model-pipeline/src/1_pop_data/query_meta_keys.py)). I think @edarin may have now updated that file on the network drive to exclude those oblasts.

doug-leasure commented:
@alaparicioc , I have now rerun the model after re-creating data from the entire pipeline using fresh seeds for the random number generators. I got the same results as I showed above. 

48 parameters slightly exceeded the rhat threshold for convergence. This was one of the worst offenders: 
![image](https://github.com/user-attachments/assets/257a55f1-37a3-40bb-9379-e4b6f0a6a9b0)

alaparicioc commented:
Getting rid of "Luhansk, Donetsk, Crimea, and Sevastopol" makes the model to converge.

doug-leasure commented:
> Getting rid of "Luhansk, Donetsk, Crimea, and Sevastopol" makes the model to converge.

@alaparicioc  That's good to know.  Can I see your results when you removed those oblasts?

alaparicioc commented:
The results are on the shared folder "~\DemSci\projects\2023_WHO_Ukraine_Population\tmp\github\UkraineNowPop-66-model-pipeline\out\modelling"
